cou/apps/base.py

```diff
@@ -793,3 +791,28 @@ def _check_application_target(self, target: OpenStackRelease) -> None:
                 f"Application '{self.name}' already configured for release equal to or greater "
                 f"than {target}. Ignoring."
             )
+
+    def _check_mismatched_versions(self, units: Optional[list[COUUnit]]) -> None:
+        """Check that there are no mismatched versions on app units.
+
+        If units are passed that means that the application will upgrade unit by unit,
+        so there are no mismatch.
```

Comment:
I thought that we had another discussion yesterday, if we have unit by unit upgrade we will check those units and not raise an exception, but just warning.

Now I realized that the warning is maybe not good idea, since NovaCompute will be always upgrade unit by unit, so we should check all units provided. I know that, like this, a user can hit an issue when trying to upgrade AZ-xyz where one unit is already upgraded. The question is what should happen if I do
```bash
cou upgrade hypervisors --az az-xyz
```
and 1 of 3 nova-compute unit is in different version. This can happen when a user previously targeted only a single machine or when other has no-empty hypervisors.

Reply:
IMO, the responsibility of selecting a unit or not depending on the target passed should happen in the [filter_hypervisors_machines](https://github.com/canonical/charmed-openstack-upgrader/blob/dev/data-plane/cou/steps/plan.py#L432). There we can have a logging saying that unit/machine x was not selected because it's already upgraded. What do you think?

Reply:
What exactly you want to move there? The logic of skipping nova-compute with different version? What about cinder?

Reply:
When you upgrade one unit the other that is collocated also gets upgraded, but I see your point...
Whatever method we implement to not generate plan for units that are already upgraded, I think it's a separate thing from this mismatched check. I don't see value on checking this on applications that upgrade unit by unit because it's an expected thing to happen.